kubernetes-simulator/simulator
# Progress state management

## Intro and reviewing notes

This ended up being a significant piece of work, I couldn't see a way to submit
a pull request any smaller than this as it's all one logical feature. I have
however gone to lengths to ensure that all the commits are neatly rebased into
discrete steps.  You will find it easier to review the code if you step through
the commits one by one looking at the changes.  Be aware though that you will
find I incrementally changed things (including the design) as I learnt more
while building, so please defer commenting until you have seen the whole set of
changes :)  A large number of the changes are an automated migration of all the
`tasks.yml` files adding a `name` property.  You can skip over these, I'm confident
the migration worked correctly - `jq` / `yq` has reformatted the files a little making it
look like there are more changes than there are.


### Features

* Introduces an abstraction for persisting the users progress on scenarios
* Includes a default implementation that saves and reads a `progress.json` file
in the `~/.kubesim` folder
* Adds functionality for the scenario tools that update the user's progress to
ask for the user's progress and send the progress back to the launch container
for it to be saved.  I.e. `start_task`, `end_task`, `next_hint`
* We now record the user's progress for all scenarios, not just the current
scenario
* The user's record of progress now persists across cluster destruction and
recreation
* Significantly improves logging of the SSH connection and interactions with
the progress state handler.  You can now debug the SSH connection, progress
persistence and the HTTP communication by looking at the logs in
`~/.kubesim/ssh.log`

### Bugfixes

* `start_task` / `end_task` command detection (these are processed by the same bin js file - it was broken)
* Don't allow re-scoring of a task that has already been scored

### Implementation details

 The change muxes an SSH reverse tunnel over the `simulator ssh attack`
 connection opening a remote port (51432) for the duration of the session which
 is forwarded to a local port (51432).  The `simulator` binary runs an HTTP
 server in a go routine listening on that port accepting connections and
 persisting the progress and sending the current progress back when requested.
 This means that the docker attack container now has `--net=host` argument so
 that it can communicate back to the launch container via the bastion... I
 don't think this is a problem given the container is ephemeral (only lasts for
the duration of the `simulator ssh attack` session) and that port is not opened on 
the firewall to the outside world.

Unfortunately the way I originally designed the `progress.json` was really tricky
to unmarshal in golang... this has meant I had to do a much bigger
refactoring to change the shape of the JSON to be more explicit for golang to
easily unmarshal it.

Note that it may seem strange that I log and return errors in the path for all
the SSH / HTTP server / Progress code, this is so that the user is notified of
problems but also the logs are written to the SSH logfile for later diagnosis.
the user will not see all of the log messages because for part of the error path
their terminal will be displaying the SSH remote shell.


### Limitations and issues:

* Using HTTP makes it easy to cheat - a user could discover the port and curl
it from the launch container to fake their score / progress
* `--net=host` flag is a bit of sledgehammer but the port is not open to the
world on the bastion.  If there is a way to do this in a more finegrained manner
I'd be keen to hear suggestions
* No end-to-end test - this doesn't concern me I have good coverage at the
boundaries.  I mock the HTTP server in the unit tests for the scenario tools
and I test the HTTP handler thoroughly.  See the manual test plan below - if
end-to-end is broken it should be obvious as soon as you use the simulator
* Interacting tests causing flickering ssh key test failures, if you hit this
rerun the make target and it will work second time (this is unrelated to this 
PR - I shall fix separately #269)
* docker goss tests are failing on infra but do not casue creation to fail(this 
is unrelated to this PR - I shall fix separately #270)
* I didn't add official documentation other than this PR as it makes it easier
to cheat and I feel that we want to keep the progress json file as an internal
implementation detail.
* The progress wire format is not versioned. Should it be?

### Testing:

Below is a **complete** manual test plan for verifying this functionality

#### Setup

* I have pushed the attack container with a tag called 'progress-state-management'
* Clone this PR and destroy any infrastructure you may have with `simulator infra destroy`
* Recreate your infrastructure with 
`simulator infra create --attack-container-tag=progress-state-management`

#### Test plan

Persistence of progress across all scenarios:

* Launch a scenario e.g.`simulator scenario launch etcd-inverted-wedge`
* Logon to your cluster with `simulator ssh attack`
* Start a task `start_task 1`
* Use all the scenario tools and check they all still work as before:
  * `next_hint` repeatedly
  * `show_hints`
  * `end_task` (choose to be scored and not to be scored)
  * `start_task` with a different task id
* exit the cluster
* Check the `~/.kubesim/progress.json` the state should reflect what you did
in the cluster
* Destroy your cluster with `simulator infra destroy`
* Create a new cluster with `simulator infra create --attack-container-tag=progress-state-management`
* Launch a different scenario e.g. `simulator scenario launch node-shock-tactics`
* Interact with the scenario tools to record some progress (see above)
* exit the attack container and check your `~/.kubesim/progress.json` is correct
* Destroy your infra with `simulator infra destroy`
* Create a new cluster `simulator infra create --attack-container-tag=progress-state-management`
* Launch the same scenario you started the first time e.g.`simulator scenario launch etcd-inverted-wedge`
* Logon to the cluster with `simulator ssh attack`
* Run `show_hints` check your progress from the last time you launched the scenario
is the same
* Review `~/.kubesim/ssh.log` and check that it appears as you would expect.
  * Is there anything superfluous or missing?
  * Could the messages be improved?
  * Are there any fields you would add or remove from the structured logging?
